service/history/timerQueueProcessorBase.go

```diff
@@ -632,50 +634,94 @@ func (t *timerQueueProcessorBase) deleteWorkflow(task *persistence.TimerTaskInfo
-			tag.TaskType(task.GetTaskType()))
+	executionStats, err := workflowContext.loadExecutionStats()
+	if err != nil {
 		return err
 	}
+	archiveInline := executionStats.HistorySize < int64(t.config.TimerProcessorHistoryArchivalSizeLimit())
+	taggedLogger := t.logger.WithTags(
+		tag.Error(err),
+		tag.WorkflowID(task.WorkflowID),
+		tag.WorkflowRunID(task.RunID),
+		tag.WorkflowDomainID(task.DomainID),
+		tag.ShardID(t.shard.GetShardID()),
+		tag.TaskID(task.GetTaskID()),
+		tag.FailoverVersion(task.GetVersion()),
+		tag.TaskType(task.GetTaskType()),
+	)
+	if !archiveInline {
+		req := &archiver.ArchiveRequest{
+			ShardID:              t.shard.GetShardID(),
+			DomainID:             task.DomainID,
+			DomainName:           domainCacheEntry.GetInfo().Name,
+			WorkflowID:           task.WorkflowID,
+			RunID:                task.RunID,
+			EventStoreVersion:    msBuilder.GetEventStoreVersion(),
+			BranchToken:          msBuilder.GetCurrentBranch(),
+			NextEventID:          msBuilder.GetNextEventID(),
+			CloseFailoverVersion: msBuilder.GetLastWriteVersion(),
+			URI:                  domainCacheEntry.GetConfig().HistoryArchivalURI,
+		}
+
+		// send signal before deleting mutable state to make sure archival is idempotent
+		if err := t.historyService.archivalClient.Archive(req); err != nil {
+			taggedLogger.Error("failed to initiate archival")
+			t.metricsClient.IncCounter(metrics.HistoryProcessDeleteHistoryEventScope, metrics.WorkflowCleanupSendArchiveSignalFailedCount)
+			return err
+		}
+	} else {
+		URI := domainCacheEntry.GetConfig().HistoryArchivalURI
+		scheme, err := common.GetArchivalScheme(URI)
+		if err != nil {
+			return err
+		}
+		historyArchiver, err := t.historyService.archiverProvider.GetHistoryArchiver(scheme, common.HistoryServiceName)
+		if err != nil {
+			return err
+		}
+		req := &carchiver.ArchiveHistoryRequest{
+			ShardID:              t.shard.GetShardID(),
+			DomainID:             task.DomainID,
+			DomainName:           domainCacheEntry.GetInfo().Name,
+			WorkflowID:           task.WorkflowID,
+			RunID:                task.RunID,
+			EventStoreVersion:    msBuilder.GetEventStoreVersion(),
+			BranchToken:          msBuilder.GetCurrentBranch(),
+			NextEventID:          msBuilder.GetNextEventID(),
+			CloseFailoverVersion: msBuilder.GetLastWriteVersion(),
+		}
+		ctxWithTimeout, cancel := context.WithTimeout(context.Background(), t.config.TimerProcessorHistoryArchivalTimeLimit())
+		defer cancel()
+		if err := historyArchiver.Archive(ctxWithTimeout, URI, req); err != nil {
+			t.metricsClient.IncCounter(metrics.HistoryProcessDeleteHistoryEventScope, metrics.WorkflowCleanupArchiveFailedCount)
+			taggedLogger.Error("failed to perform workflow history archival inside timer queue processor")
+			return err
+		}
+	}
+
 	if err := t.deleteCurrentWorkflowExecution(task); err != nil {
 		return err
 	}
 	if err := t.deleteWorkflowExecution(task); err != nil {
 		return err
 	}
+	if archiveInline {
```

Comment:
In order to handle this in the model where client does both cases just return a result from client which is a struct including a field describing how archival was done.

Reply:
I think pass `archiveInline` to the client.Archive() method makes more sense. Basically, you give the archive client a request and how tell it how archive should be done. Making the client figure out how archive should be done on its own seems the wrong level of abstraction to me. Also `workflowContext` and the dynamicconfig need to be passed to the `Archive()` method to support which I think is weird.